src/plugins/dataflow/components/ui/dataflow-program-topbar.tsx

```diff
@@ -86,31 +65,237 @@ export const DataflowProgramTopbar = (props: TopbarProps) => {
 
   return (
     <div className="program-editor-topbar">
-      <div className="topbar-left">
-        {<IconButton
-          icon="serial"
-          key="serial"
-          onClickButton={props.onSerialRefreshDevices}
-          title="Refresh Serial Connection"
-          disabled={props.readOnly}
-          className={classes}
-        />}
-        <div className="serial-message">
-          { serialMessage() }
+
+        <div className="topbar-icon">
+          {<IconButton
+            icon="serial"
+            key="serial"
+            onClickButton={onSerialRefreshDevices}
+            title="Refresh Serial Connection"
+            disabled={readOnly}
+            className={classes}
+          />}
+          <div className="serial-message">
+            { serialMessage() }
+          </div>
         </div>
+
+        <div className="topbar-center-container">
+          <div className="topbar-blank-or-play">
+            {
+              (programRecordState === 1 || programRecordState === 2) &&
+              <PlaybackButton programRecordState={programRecordState}/>
+            }
+          </div>
+          <RateSelectorOrPlayBack
+            rateOptions={programDataRates}
+            dataRate={dataRate}
+            onRateSelectClick={onRateSelectClick}
+            readOnly={readOnly}
+            programRecordState={programRecordState}
+            numNodes={numNodes}
+            onRecordDataChange={onRecordDataChange}
+            finished={finished}
+            handleFinished={handleFinished}
+          />
+          <RecordStopOrClearButton
+            programRecordState={programRecordState}
+            buttonClickHandler={onRecordDataChange}
+          />
+        </div>
+
+        <div className="topbar-right">
+          {props.showRateUI && <span className={"rate-ui"}>{`${props.lastIntervalDuration}ms`}</span>}
+        </div>
+
+    </div>
+  );
+};
+
+/* ==[ Sampling Rate Selector ] == */
+interface IRateSelectorProps {
+  rateOptions: ProgramDataRate[];
+  dataRate: number;
+  onRateSelectClick: (rate: number) => void;
+  readOnly: boolean;
+  programRecordState: number;
+  numNodes: number
+  finished: boolean;
+  handleFinished: (isFinished: boolean)=> void;
+  onRecordDataChange: () => void;
+}
+
+function formatTime(seconds: number) {
+  const minutes = Math.floor(seconds / 60);
+  const remainingSeconds = seconds % 60;
+  const formattedMinutes = minutes.toString().padStart(3, '0');
+  const formattedSeconds = remainingSeconds.toString().padStart(2, '0');
+  return `${formattedMinutes}:${formattedSeconds}`;
+}
+
+const RateSelectorOrPlayBack = (props: IRateSelectorProps) => {
+  const { onRateSelectClick, readOnly, dataRate, rateOptions, programRecordState,
+          numNodes, handleFinished, onRecordDataChange} = props;
+
+  /* ==[ Total Recording Time  - Calculate] format as "MMM:SS" */
+  const totalTimeSec = Math.floor((dataRate / 1000) * (totalSamples/numNodes));
```

Comment:
I don't think we need to worry about the number of nodes when calculating how long we can record for.  Using this calculation, the recording time is brought down significantly with every new node. 

I believe you could leave it at:
```
const totalTimeSec = Math.floor(dataRate/1000) * totalSamples))
```

I tried this, and it gave me the expected recording times for each datarate.

Reply:
^^^ @dennisrcao, @lbondaryk reminded me that we are restricting this by number of recorded points, and not by number of recorded frames.  So, leave the calc in place!